Updates for compatibility with django 4.0+

rlskoeser opened the pull request.

acdha commented on requirements.txt:
```diff
@@ -1,2 +1,2 @@
-django
+django==4.1
```
This would prevent the install of a newer version. I typically would set the minimum version and then go up – for example `django>=4.1` or, if we wanted to restrict it to the ones we had in mind while testing, perhaps `django>=4.1,<6`.

rlskoeser replied:
Thanks, I will switch to that.

I think I set it as a shortcut for testing with tox. I was trying to figure out how to do a matrix build with tox to test multiple versions of django but I'm not that familiar with tox and I use github actions for matrix builds

acdha replied:
Yeah, I'd like to pivot to GitHub actions for this as well since it'd be faster but I no longer work on the project where I primarily used this code so it's somewhat far down the to-do list.

rlskoeser replied:
if I have energy to make a quick stab at github actions based on my other projects would it be ok to include in this pr?

acdha commented on tox.ini:
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py27, py35, py36
+envlist = py39
```
Python 3.9 is pretty old at this point and hit the end of support 2 years ago, although it's common on the LTS Linux distributions. I'd go up to 3.12 at this point.

```suggestion
envlist = py39,py310,py311,py312
```

rlskoeser commented:
@acdha I've simplified the conditional imports, made the changes you flagged, marked the unit tests that are failing for me to be skipped (I can't tell if they are expected to pass), and added a basic github actions unit test with a matrix build. Do you want to look it over and approve the workflow if it looks ok?

**update**: tests are passing on github actions on my branch https://github.com/rlskoeser/django-tabular-export/actions/runs/10410420372  

LMK if you want me to adjust the build matrix and drop the older versions of django before you approve it to run in your org

acdha commented on tests/test_tabular_exporter.py:
```diff
@@ -159,6 +159,7 @@ def test_export_to_csv_response(self):
                                    '1,2\r\n', '3,4\r\n', 'abc,def\r\n',
                                    '2015-08-28T00:00:00,2015-08-28\r\n'])
 
+    @unittest.skip("failing - does not return an HttpResponse")
```
I was not expecting that…

rlskoeser replied:
yeah, I wasn't sure how to troubleshoot this - I skipped it for now so I could make sure the build was working

could it be something I don't have installed? or something that changed with debug behavior?

acdha replied:
Yeah, I'm waiting for the punchline to be that it's returning a redirect or something like that.

rlskoeser replied:
there is a redirect but when I added the arg to follow it rendered an admin list template which I didn't think was the expected behavior (based on my memory from the other day)

acdha replied:
Ah, running it without the skip:

```
TypeError: never_cache didn't receive an HttpRequest. If you are decorating a classmethod, be sure to use @method_decorator.```

acdha replied:
Okay, so it looks like this was a backwards-incompatible change in Django but there's an easy fix to switch to `add_never_cache_headers` which is simpler anyway.

acdha commented:
Those fast matrix builds are so nice, thank you for adding them!

acdha commented on tests/test_admin_actions.py:
```diff
@@ -64,10 +68,11 @@ def test_export_to_csv_action(self):
+        self.assertRegex(content[1], r'^1,TEST ITEM 1,0\r\n')
+        self.assertRegex(content[2], r'^2,TEST ITEM 2,0\r\n')
 
+    @unittest.skip("error (where is custom export configured?)")
```
The underlying problem here is probably that the form parameters which the Django admin interface expects has probably changed: it's generating a 302 redirect to the requested URL so the test code must be missing something basic.

```python
(Pdb) p response
<HttpResponseRedirect status_code=302, "text/html; charset=utf-8", url="/admin/tests/testmodel/">
(Pdb) p changelist_url
'/admin/tests/testmodel/'
```

blms replied:
Hi both! @rlskoeser asked me to take a look at this.

The error message Django sends back here is "No action selected." You can check this with:
```py
from django.contrib.messages import get_messages
messages = [m.message for m in list(get_messages(response.wsgi_request))]
```
I notice that there's no action actually named `custom_export_to_csv_action` so maybe this is the intended behavior? While that is the same error message you get on form validation errors, I don't see any difference in the form params—the underlying Django code hasn't actually changed ([Django 3.2.x](https://github.com/django/django/blob/3591e1c1acbd7c13174275367c3fdf012cb0413b/django/contrib/admin/options.py#L1384-L1420), [Django 5](https://github.com/django/django/blob/01a4d8a3c741b3129d481ef3515084a199d21222/django/contrib/admin/options.py#L1640-L1678)), and I can't get this test to succeed under Django 3.2 either.

It seems correct to redirect with this error message unless an admin action with that exact name was registered, unless I am missing something.

rlskoeser replied:
Thanks for looking into this, @blms. 

@acdha let us know how you'd like to resolve this.

acdha replied:
Looking at aa890c8d44afc63b835459637fa6e99b577b5d2b, I do think that test might simply never have worked. It'd be nice to have one but the path of least resistance is probably just removing it now.

rlskoeser commented:
Hello, @acdha, hope things are going ok for you and LoC folks these days.  

I'm circling back to see if we can get this update merged so you can release a new version released, or if we should start looking at alternate packages.  In the long run, do you recommend looking for an alternative?

In the short term: both @blms and I looked at the failing test and don't see any evidence that it ever worked. Can we remove the failing test or do you need a test case for the custom export configuration?

blms commented:
Thanks @acdha for pointing towards the original commit. I see now the intended behavior of this test, and what's missing is just a bit of configuration in `tests/admin.py`'s `TestModelAdmin` class:

https://github.com/rlskoeser/django-tabular-export/blob/db0a0c5062236d18b751f62f7b34532ded472b1c/tests/admin.py#L12-L13

If you add this custom action to the test admin, the original test will pass as written:

```py
class TestModelAdmin(admin.ModelAdmin):
    def custom_export_to_csv_action(self, request, queryset):
        # Add a custom action with the extra verbose name "number of tags"
        return export_to_csv_action(
            self,
            request,
            queryset,
            extra_verbose_names={'tags_count': 'number of tags'},
        )

    actions = (export_to_excel_action, export_to_csv_action, custom_export_to_csv_action)
```

@rlskoeser let me know if you'd prefer a separate PR, but I think this is simple enough to just drop into yours and remove the skip.

rlskoeser commented:
Thanks, @blms !

@acdha I've pushed an update that includes Ben's fix and some minor tweaks to get the test passing.  I went ahead and added python 3.13 and noticed that the classifiers in the setup file were out of date with what's actually being tested, so I updated them as well.

Hoping this gets the project to where you can easily merge and release a new version.

rlskoeser commented:
Hi, @acdha. I'm sure you have plenty of other things going on; following up to see if you can get this merged and a new version released. We'd appreciate it!
